File: android/src/main/java/com/visioncameramlkit/infrastructure/serializers/TextRecognitionSerializer.kt
package com.visioncameramlkit.infrastructure.serializers

import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.visioncameramlkit.domain.models.BoundingBox
import com.visioncameramlkit.domain.models.Corner
import com.visioncameramlkit.domain.models.TextBlock
import com.visioncameramlkit.domain.models.TextElement
import com.visioncameramlkit.domain.models.TextLine
import com.visioncameramlkit.domain.models.TextRecognitionResult
import com.visioncameramlkit.domain.models.TextSymbol

@Suppress("TooManyFunctions")
object TextRecognitionSerializer {
  fun toWritableMap(result: TextRecognitionResult): WritableNativeMap =
    WritableNativeMap().apply {
      putString("text", result.text)
      putArray("blocks", toBlocksArray(result.blocks))
    }

  fun toReactNativeMap(result: TextRecognitionResult): HashMap<String, Any> {
    @Suppress("UNCHECKED_CAST")
    return toWritableMap(result).toHashMap() as HashMap<String, Any>
  }

  private fun toBlocksArray(blocks: List<TextBlock>): WritableNativeArray =
    WritableNativeArray().apply {
      blocks.forEach { block -> pushMap(toBlockMap(block)) }
    }

  private fun toBlockMap(block: TextBlock): WritableNativeMap =
    WritableNativeMap().apply {
      putString("text", block.text)
      block.bounds?.let { putMap("bounds", toBoundsMap(it)) }
      block.corners?.let { putArray("corners", toCornersArray(it)) }
      putArray("lines", toLinesArray(block.lines))
      putArray("languages", toLanguagesArray(block.languages))
    }

  private fun toLinesArray(lines: List<TextLine>): WritableNativeArray =
    WritableNativeArray().apply {
      lines.forEach { line -> pushMap(toLineMap(line)) }
    }

  private fun toLineMap(line: TextLine): WritableNativeMap =
    WritableNativeMap().apply {
      putString("text", line.text)
      line.bounds?.let { putMap("bounds", toBoundsMap(it)) }
      line.corners?.let { putArray("corners", toCornersArray(it)) }
      putArray("elements", toElementsArray(line.elements))
      line.confidence?.let { putDouble("confidence", it) }
      line.angle?.let { putDouble("angle", it) }
      putArray("languages", toLanguagesArray(line.languages))
    }

  private fun toElementsArray(elements: List<TextElement>): WritableNativeArray =
    WritableNativeArray().apply {
      elements.forEach { element -> pushMap(toElementMap(element)) }
    }

  private fun toElementMap(element: TextElement): WritableNativeMap =
    WritableNativeMap().apply {
      putString("text", element.text)
      element.bounds?.let { putMap("bounds", toBoundsMap(it)) }
      element.corners?.let { putArray("corners", toCornersArray(it)) }
      putArray("symbols", toSymbolsArray(element.symbols))
      element.confidence?.let { putDouble("confidence", it) }
      element.angle?.let { putDouble("angle", it) }
      putArray("languages", toLanguagesArray(element.languages))
    }

  private fun toSymbolsArray(symbols: List<TextSymbol>): WritableNativeArray =
    WritableNativeArray().apply {
      symbols.forEach { symbol -> pushMap(toSymbolMap(symbol)) }
    }

  private fun toSymbolMap(symbol: TextSymbol): WritableNativeMap =
    WritableNativeMap().apply {
      putString("text", symbol.text)
      symbol.bounds?.let { putMap("bounds", toBoundsMap(it)) }
      symbol.corners?.let { putArray("corners", toCornersArray(it)) }
      symbol.confidence?.let { putDouble("confidence", it) }
      symbol.angle?.let { putDouble("angle", it) }
      putArray("languages", toLanguagesArray(symbol.languages))
    }

  private fun toBoundsMap(bounds: BoundingBox): WritableNativeMap =
    WritableNativeMap().apply {
      putDouble("x", bounds.x)
      putDouble("y", bounds.y)
      putDouble("centerX", bounds.centerX)
      putDouble("centerY", bounds.centerY)
      putDouble("width", bounds.width)
      putDouble("height", bounds.height)
      putDouble("top", bounds.top)
      putDouble("left", bounds.left)
      putDouble("bottom", bounds.bottom)
      putDouble("right", bounds.right)
    }

  private fun toCornersArray(corners: List<Corner>): WritableNativeArray =
    WritableNativeArray().apply {
      corners.forEach { corner -> pushMap(toCornerMap(corner)) }
    }

  private fun toCornerMap(corner: Corner): WritableNativeMap =
    WritableNativeMap().apply {
      putDouble("x", corner.x)
      putDouble("y", corner.y)
    }

  private fun toLanguagesArray(languages: List<String>): WritableNativeArray =
    WritableNativeArray().apply {
      languages.forEach { pushString(it) }
    }
}

File: ios/Infrastructure/Serializers/TextRecognitionSerializer.swift
import Foundation

class TextRecognitionSerializer {

  static func toReactNativeMap(_ result: TextRecognitionResult) -> [String: Any] {
    var map = [String: Any]()
    map["text"] = result.text
    map["blocks"] = toBlocksArray(result.blocks)
    return map
  }

  private static func toBlocksArray(_ blocks: [TextBlock]) -> [[String: Any]] {
    return blocks.map { toBlockMap($0) }
  }

  private static func toBlockMap(_ block: TextBlock) -> [String: Any] {
    var map = [String: Any]()
    map["text"] = block.text
    if let bounds = block.bounds {
      map["bounds"] = toBoundsMap(bounds)
    }
    if let corners = block.corners {
      map["corners"] = toCornersArray(corners)
    }
    map["lines"] = toLinesArray(block.lines)
    map["languages"] = block.languages
    return map
  }

  private static func toLinesArray(_ lines: [TextLine]) -> [[String: Any]] {
    return lines.map { toLineMap($0) }
  }

  private static func toLineMap(_ line: TextLine) -> [String: Any] {
    var map = [String: Any]()
    map["text"] = line.text
    if let bounds = line.bounds {
      map["bounds"] = toBoundsMap(bounds)
    }
    if let corners = line.corners {
      map["corners"] = toCornersArray(corners)
    }
    map["elements"] = toElementsArray(line.elements)
    map["confidence"] = line.confidence
    map["angle"] = line.angle
    map["languages"] = line.languages
    return map
  }

  private static func toElementsArray(_ elements: [TextElement]) -> [[String:
    Any]]
  {
    return elements.map { toElementMap($0) }
  }

  private static func toElementMap(_ element: TextElement) -> [String: Any] {
    var map = [String: Any]()
    map["text"] = element.text
    if let bounds = element.bounds {
      map["bounds"] = toBoundsMap(bounds)
    }
    if let corners = element.corners {
      map["corners"] = toCornersArray(corners)
    }
    map["symbols"] = element.symbols.map { toSymbolMap($0) }
    map["confidence"] = element.confidence
    map["angle"] = element.angle
    map["languages"] = element.languages
    return map
  }

  private static func toSymbolMap(_ symbol: TextSymbol) -> [String: Any] {
    var map = [String: Any]()
    map["text"] = symbol.text
    if let bounds = symbol.bounds {
      map["bounds"] = toBoundsMap(bounds)
    }
    if let corners = symbol.corners {
      map["corners"] = toCornersArray(corners)
    }
    map["confidence"] = symbol.confidence
    map["angle"] = symbol.angle
    map["languages"] = symbol.languages
    return map
  }

  private static func toBoundsMap(_ bounds: BoundingBox) -> [String: Double] {
    return [
      "x": bounds.x,
      "y": bounds.y,
      "centerX": bounds.centerX,
      "centerY": bounds.centerY,
      "width": bounds.width,
      "height": bounds.height,
      "top": bounds.top,
      "left": bounds.left,
      "bottom": bounds.bottom,
      "right": bounds.right,
    ]
  }

  private static func toCornersArray(_ corners: [Corner]) -> [[String: Double]] {
    return corners.map { toCornerMap($0) }
  }

  private static func toCornerMap(_ corner: Corner) -> [String: Double] {
    return [
      "x": corner.x,
      "y": corner.y,
    ]
  }
}
